### ISSUE TITLE
Unhandled Sensor: Leak Detector

### ISSUE THREAD

**Issue by eafarris** [garbled1/homeassistant_ecowitt#25]
The GW1000 supports up to four water leak detectors (WH55). After adding one to the gateway, Home Assistant reports:

```
2020-10-19 20:34:18 WARNING (MainThread) [custom_components.ecowitt] Unhandled sensor type leak_ch1
2020-10-19 20:34:18 WARNING (MainThread) [custom_components.ecowitt] Unhandled sensor type leakbatt1
```

**Comment by eafarris** [comment]
Further log entries that are probably useful:

```
2020-10-19 20:36:19 WARNING (MainThread) [custom_components.ecowitt] Unhandled sensor type leak_ch1 value 0, file a PR.
2020-10-19 20:36:19 WARNING (MainThread) [custom_components.ecowitt] Unhandled sensor type leakbatt1 value 5, file a PR.
```

**Comment by garbled1** [comment]
Hrmm.. I'm guessing this is a binary sensor.. let me think a bit.. this is slightly more difficult to handle.

**Comment by garbled1** [comment]
In theory, this should work in 0.3.2 (and HA 0.115 or >)  Please test and let me know.

**Comment by eafarris** [comment]
It does show up in HA now, but its state is 'unknown,' and triggering the detector does not change its state. Is there a way I can see the raw data from the gateway and send it to you?

The battery shows up, too, as "5". I assume that's a power level (out of 5). Is that correct?

**Comment by garbled1** [comment]
Yeah, I haven't quite nailed down the batteries yet.  Apparently they are different for every sensor, so I don't have a good setup for them.  I think for your device, it's a range of 0-5.

so.. one thing you could do, that works reasonably well if you have shell access and root:

tcpdump -A port 4199

This should dump the raw data coming from the ecowitt out, and we should be able to see the data as it flows.. the output is a little ugly, but you should be able to find the incoming json..

You will get a bunch of garbage, but somewhere in there you should see:

```
10:10:28.369778 IP XX.3777 > XX.4199: Flags [P.], seq 1:579, ack 1, win 5840, length 578
E..j.(.....:......
K...g.n....W.P...7...POST /data/report/ HTTP/1.1
HOST: XX
Connection: Close
Content-Type: application/x-www-form-urlencoded
Content-Length: 431

PASSKEY=XXX&stationtype=EasyWeatherV1.4.9&dateutc=2020-11-13+17:10:24&tempinf=76.8&humidityin=22&baromrelin=28.760&baromabsin=28.760&tempf=65.8&humidity=24&winddir=319&windspeedmph=0.9&windgustmph=1.1&rainratein=0.000&eventrainin=0.000&dailyrainin=0.000&weeklyrainin=0.024&monthlyrainin=0.028&yearlyrainin=0.843&solarradiation=375.53&uv=3&pm25_ch1=8.0&pm25_avg_24h_ch1=5.2&freq=915M&model=HP3500_V1.6.2
```

It would be useful to see a trigger and non-trigger.. Maybe I have the code that detects it wrong...  Like does it only send updates if the state changes, or does it send continual updates of "ok" ?

**Comment by garbled1** [comment]
OK, I've been investigating this.. It's definitely my code.  Working on a fix.. might be a little bit..  it's a bit confusing in there.

**Comment by garbled1** [comment]
This fix will be in 0.5 when I release it in hopefully a week or two.  I have it fixed in my branch, but that is not ready for primetime at all, and it's a bit too complex to backport.

**Comment by eafarris** [comment]
Sorry for the delay. I'm running hassio on a Proxmox VM, so, while I do have root, it's pretty pared down (no tcpdump). So I fired up a Ubuntu VM and a SimpleHTTPServer and ran Wireshark against it. The data coming off the GW1000 to the "server" does have "leak_ch1" = "0" and then "leak_ch1" = "1" (when a leak is detected) as form data in a POST, as you indicated above. The "leak_ch1" is _always_ in the form data, set to "0" if no leak is detected. It does seem to wait its normal interval before reporting a "1".

Now that I have this setup, I can do some more testing if you need. Thanks for all your hard work on this.

**Comment by garbled1** [comment]
NP.  in a different issue someone just happened to have a dump of the leak sensor, so I pasted it into my fake client and got a working simulator.  The config_flow branch now has it completely functional.

I'm hoping to do a little more testing of that branch, and then release 0.5 soon..  As of this morning I consider that branch to be stable, if anyone out there is feeling like being a beta tester.

**Comment by garbled1** [comment]
OK, this should be fixed in the 0.5 release.  Available now from your local HACS store!

**Comment by eafarris** [comment]
Confirmed. Thanks so much!
